_posts/2023-08-29-kafka-node-pools-kraft.md

```diff
@@ -0,0 +1,177 @@
+---
+layout: post
+title: "Kafka Node Pools: Supporting KRaft (ZooKeeper-less Apache Kafka)"
+date: 2023-08-29
+author: jakub_scholz
+---
+
+The basic architecture of an Apache Kafka cluster with Apache ZooKeeper is pretty straightforward.
+You have a ZooKeeper cluster with one or more ZooKeeper nodes - typically using an odd number of nodes, most often 3.
+Then you have the Kafka cluster with one or more Apache Kafka brokers that is bootstrapped using the ZooKeeper cluster.
+And that's it.
+There aren't really any other options
+But when we remove ZooKeeper to introduce the KRaft (Kafka Raft metadata) mode, the architectural landscape changes significantly.
+In this blog post, we will look at the various KRaft architectures and how Strimzi can support them using the Kafka node pools.
+
+<!--more-->
+
+_This post is part of a series about Kafka node pools.
+Other posts include:_
+
+* _[Part 1 - Introduction](https://strimzi.io/blog/2023/08/14/kafka-node-pools-introduction/)_
+* _[Part 2 - Node ID Management](https://strimzi.io/blog/2023/08/23/kafka-node-pools-node-id-management/)_
+* _[Part 3 - Storage & Scheduling](https://strimzi.io/blog/2023/08/28/kafka-node-pools-storage-and-scheduling/)_
+* _Part 4 - KRaft (this post)_
+
+### KRaft architectures
+
+In the KRaft mode, the Kafka cluster runs without an Apache ZooKeeper cluster as a dependency.
+The Kafka nodes in the KRaft mode have _controller_ and/or _broker_ roles.
+The nodes with the _controller_ role are responsible for maintaining the Kafka metadata, handling leader elections, and helping to bootstrap the Kafka cluster.
+They assume the responsibilities previously handled by the ZooKeeper nodes.
+The nodes with the _broker_ role have the same responsibilities as the Kafka brokers in the ZooKeeper-based Kafka cluster.
+They receive the messages from the producers, store them, and pass them to the consumers.
+One node can also have both roles - a controller and a broker -  at the same time.
+This offers a number of configuration options for your KRaft-based Kafka cluster.
+
+You might start with a small Kafka cluster consisting only of combined nodes that all have controller and broker roles.
+Typically, it would have three nodes.
+But if needed you can also run _a whole Kafka cluster_ in a single node.
+Such a cluster is suitable for example for development and testing.
+Or - in some cases - it might be even considered for a small production cluster.
+
+![Three-node Kafka cluster with combined nodes](/assets/images/posts/2023-08-29-kafka-node-pools-kraft-combined-nodes.png)
+
+For a big production cluster, you would normally use dedicated controller nodes and dedicated broker nodes.
```

Comment:
I'm open to suggestions on how to make this less _controversial_ if you feel that is how it looks. What is the thing which is the problem? That the combined nodes `might be considered for a small production cluster`? Or the `big production cluster` formulation? The goal is not to provide definitive guidance on when to use what. So I think that as long as the blog post sufficiently explains why are all these modes important, it does not need to be controversial.

Would this work for you? (***bold*** parts are new/changed, ~~strikethrough~~ will be removed)

----

You might start with a small Kafka cluster consisting only of combined nodes that all have controller and broker roles.
Typically, it would have three nodes.
But if needed you can also run _a whole Kafka cluster_ in a single node.
Such a cluster is suitable for example for development and testing.
***It might be also useful in other resource-constraint environments.
For example when running Kafka at the edge.***
~~Or - in some cases - it might be even considered for a small production cluster.~~

![Three-node Kafka cluster with combined nodes](/assets/images/posts/2023-08-29-kafka-node-pools-kraft-combined-nodes.png)

***For a production cluster, you would normally use dedicated controller nodes and dedicated broker nodes.***
You would typically maintain three controller nodes for availability, and add as many broker nodes as you need.
***This is also the architecture to which you would migrate from a ZooKeeper-based cluster.***

![Three-node Kafka cluster with dedicated controller and broker nodes](/assets/images/posts/2023-08-29-kafka-node-pools-kraft-separate-nodes.png)

And somewhere in between is an architecture where some nodes have both controller and broker roles and some have only the broker role.
~~It is a good fit for example when you run on bare metal and have only powerful worker nodes that you want to be fully utilized and that would be too big to run only as controllers.~~
It is also an intermediate step when migrating between the previous two architectures. 

----

That said, my feeling is that the examples of when it might be used I use there are realistic, in line with the expectations of many users, and correspond to how users will expect to use them. Blindly insisting on dedicated controller nodes for every production cluster will be very shortsighted and will not do any good for Kafka. But no problem to leave this out of this blog post.

Reply:
Thanks, I like your suggested construction much better! @mimaison @showuon are you also happy with this?

Reply:
Looks good. Thanks.

Reply:
Ok, I updated the PR @tombentley